### ISSUE TITLE
opam does not update git branches

### ISSUE THREAD

**Issue by RalfJung** [ocaml/opam#3727]
On our CI, we (as of now) are using the following commands to update all installed packages and, if needed, install the package defined in `build-deps/` (that directory contains an `opam` file):
```
opam init --no-setup --disable-sandboxing --reinit -y
opam update
opam upgrade -y --all build-dep/
```
This used to work fine (not always with the same command sequence), but stopped reliably updating packages sometime around Dec 19th.  You can see this, for example, in [this CI log](https://gitlab.mpi-sws.org/FP/iris-coq/-/jobs/22121):
```
[prepare-opam] Done! Coq version we are going to use:
The Coq Proof Assistant, version 8.10+alpha (December 2018)
compiled on Dec 18 2018 15:23:34 with OCaml 4.07.0
version     dev
source-hash 4c733a92
```
That source-hash is outdated, there have been new commits in the branch that version "dev" points to since then.

On Dec 19th, we changed two things: we updated to opam 2.0.2, and we changed the command sequence from
```
opam update
opam upgrade -y --all build-dep/
```
to
```
opam init --no-setup --disable-sandboxing --reinit -y
opam upgrade -y --all build-dep/
```
I did the `opam init` change because the opam release notes said to do this after the update, and of course this being CI I cannot just interactively run this once, I have to make sure it runs each time an old cached opam root is used.

I assumed the changed command sequence might be the issue (`opam init` might not have *all* the effects of `opam update`), so I added `opam update --development`, but that doesn't seem to update git repos (which btw I think is a bug, the docs make it sound like `opam update --repositories && opam update --development` is the same as `opam update`).  So now I added a full `opam update`, redundantly updating the repositories, but it *still* does not update coq.

Another example is [this CI log](https://gitlab.mpi-sws.org/FP/LambdaRust-coq/-/jobs/21848) which shows a build failure because coq-stdpp was not properly updated. This was using the `opam init && opam upgrade` command sequence. coq-stdpp is pinned to a git branch, and at that point the repository already contained a new function `tc_to_bool`, but opam failed to update coq-stdpp and hence the build fails because that method is missing.  We had this failure for *all* our projects, and only fixed it (worked around it, really) by changing the URL of the pinned git repository -- *that* seems to trigger the right update.

```
# opam config report
# opam-version      2.0.2 
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=debian os-version=9
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              20
# repositories      4 (http), 1 (version-controlled) (default repo at 0f6e2524)
# pinned            2 (git), 1 (rsync), 1 (version)
# current-switch    ocaml-base-compiler.4.02.3
```

**Comment by RalfJung** [comment]
I've now added `opam upgrade coq -y` to the command sequence. That "fixes" the problem, but clearly it shouldn't be necessary after a `opam upgrade -y --all $PACKAGE`.

**Comment by rjbou** [comment]
On upgrade: opam upgrades installed package, by default repo and pinned ones. If a package is not installed, but pinned, it won't be installed. As on you script you pin the packages and then you do an upgrade to install them, it won't install them all. When you use `opam upgrade -y --all build-dep/`, it installs everything because of waterfall dependencies: `coq-lambda-rust-builddep -> coq-iris -> coq-stdpp -> coq`

> I've now added `opam upgrade coq -y` to the command sequence. That "fixes" the problem, but clearly it shouldn't be necessary after a `opam upgrade -y --all $PACKAGE`.

In [this log](https://gitlab.mpi-sws.org/FP/iris-coq/-/jobs/22121), there is an
```
[NOTE] coq.dev has previously been updated with --working-dir, not resetting unless explicitly selected
```
This can be an origin of why coq doesn't update (note: I didn't reproduce it, with a fresh switch).

> the docs make it sound like `opam update --repositories && opam update --development` is the same as `opam update`

It is, if not it's a bug. Can you reproduce or still have the log of failing update?

> I did the opam init change because the opam release notes said to do this after the update

Exactly, in the [blog post](https://opam.ocaml.org/blog/opam-2-0-2/) it is recommanded to do an reinit _in order to update sandbox scripts_. As you disable sandboxing for CI, you don't need it.

> coq-stdpp is pinned to a git branch, and at that point the repository already contained a new function `tc_to_bool`, but opam failed to update coq-stdpp and hence the build fails because that method is missing

On your script, you have
```
opam update
opam pin coq-stdpp <url> -n
opam pin coq-iris <url> -n
opam upgrade --all build-dep/
```
Pinning doesn't update source directory. On your opam root these packages are already pinned to that url, so when you pin them again, it does nothing. Then you upgrade directly without updating those pinned packages. Why don't you update after pinning?
Other solutions are to unpin them before pinning, or to do an `opam update --dev` before the upgrade.

**Comment by RalfJung** [comment]
> On upgrade: opam upgrades installed package, by default repo and pinned ones. If a package is not installed, but pinned, it won't be installed.

In this case, Coq is already pinned *and* installed.  It just doesn't get upgraded.

> Can you reproduce or still have the log of failing update?

It wasn't failing, but I observed `opam update --development` to not say anything about a git repo, whereas `opam update` mentioned a git repo.  But I cannot find the log now so maybe I just misremembered.  I'll look into this later once we resolved this one.

> As you disable sandboxing for CI, you don't need it.

Well, I figured I don't know which other things might require such an update in the future, so I'd better just add this reinit anyway.

> Pinning doesn't update source directory. On your opam root these packages are already pinned to that url, so when you pin them again, it does nothing.

That's okay.

A later `opam upgrade`, however, should do something if they got changed upstream, right? That's what's not happening here.

> Why don't you update after pinning?

Uh... the order here originates from opam 1 times where we couldn't `pin -n`, we had to actually install things in that pinning loop.  What difference would that make?

**Comment by rjbou** [comment]
> In this case, Coq is already pinned and installed. It just doesn't get upgraded.

Is it the same case than described in #3651?

> A later opam upgrade, however, should do something if they got changed upstream, right?

It lead to the discussion in #3613, `upgrade` operation installs what is already present (assuming updates have been done on the `update` operation), whereas `install` operation, on pin, updates source before installing.

**Comment by RalfJung** [comment]
> Is it the same case than described in #3651?

No. There, a re-build happens but with the wrong source.

Here, no re-build happens.  I fixed this by adding `opam upgrade coq` to the end of our command sequence.  But after a `opam upgrade -y --all build-dep/`, doing `opam upgrade -y coq` really shouldn't make a difference -- I already told it to update everything!

**Comment by rjbou** [comment]
Is your process:
```
$ opam pin coq -k version -n -y
$ opam install coq -y
# update coq repository
$ opam update
$ opam pin coq -k version -n -y
$ opam upgrade -y --all build-dep/
```

**Comment by RalfJung** [comment]
No, the command sequence is always the same:
```
$ opam update
$ opam pin coq -k version -n -y
$ opam upgrade -y --all build-dep/
# Coq gets uüpdated
$ opam update
$ opam pin coq -k version -n -y
$ opam upgrade -y --all build-dep/
```

**Comment by rjbou** [comment]
In this case, version-pinned coq is update on the `opam update`, then at upgrade, new source is compiled.

But there is an old `--working-dir` that prevents coq source directory to be updated, as you can see in the update part on [log](https://gitlab.mpi-sws.org/iris/iris/-/jobs/22121). This check is done by looking at source directory, and in the case where source is _dirty_ (vc diff not empty), opam supposes that it has been updated by `--working-dir` option, and doesn't update it.

**Comment by RalfJung** [comment]
> In this case, version-pinned coq is update on the opam update, then at upgrade, new source is compiled.

Well, that's what I would hope.  But instead, at `upgrade`, nothing happens.

> But there is an old --working-dir that prevents coq source directory to be updated, as you can see in the update part on log. This check is done by looking at source directory, and in the case where source is dirty (vc diff not empty), opam supposes that it has been updated by --working-dir option, and doesn't update it.

I do not understand what this means.  Are you saying something (maybe the Coq build process) changed the sources and that confused opam?
But then why does `opam upgrade coq` ignore this problem?

**Comment by rjbou** [comment]
If you didn't use `--working-dir` previously, then yes, something changed the opam coq source directory (`~/.opam/<switch>/.opam-switch/sources/coq`).

As upgrade is not directly bound to update operation, it doesn't perform the same checks, that in this case results to an inconsistency among update restriction on dirty source and upgrade that doesn't check it all.

**Comment by RalfJung** [comment]
> If you didn't use --working-dir previously

We certainly didn't.  All that ever happened is the command sequence shown above:
```
$ opam update
$ opam pin coq -k version -n -y
$ opam pin add -y -n build-dep/
$ opam upgrade -y --all build-dep/
```

**Comment by RalfJung** [comment]
Also it would be nice (for non-interactive use) to have a way to force opam into ignoring dirty build dirs.  Otherwise this is extremely fragile, if any package screws up and modifies their sources during build then the entire CI is stuck and needs manual intervention.

**Comment by rjbou** [comment]
You can run once `opam update coq` to force update coq source directory.

> Is there a way to tell opam to check if the dir is clean? 

opam itself checks it in its process, there is not options to just check if source is clean or dirty. There is two way to do that:
- by looking directly on internal source: go to source directory and check result of `git diff`
- by using opam libraries: write a small piece of code that checks if the source is clean

> Did the behavior of these checks change with opam 2.0.2?

No, the most related change was a fix on format upgrade 1.2 to 2.0 for source of version pinned package, but I don't think it is your case. Yous can find the list of changes in this [blog post](https://opam.ocaml.org/blog/opam-2-0-2/).

**Comment by RalfJung** [comment]
> You can run once opam update coq to force update coq source directory.

Sure, but that doesn't help for CI. I'd have to manually do this for every package, or so.

**Comment by RalfJung** [comment]
In https://gitlab.mpi-sws.org/FP/gpfsl/-/jobs/23012 this bug leads to another stuck state: opam forgets to update Coq, thus compiles Iris with an old Coq version that is not compatible any more, and aborts the process.  It never even reaches the `opam upgrade coq` that I added as a hack.

Seems I have to delete the build cache of all our projects. Moving the `opam upgrade coq` up won't help, coq and Iris need to be updated together for this to work.

**Comment by RalfJung** [comment]
This just happened again, this time it's coq-stdpp not being upgraded in <https://gitlab.mpi-sws.org/iris/lambda-rust/-/jobs/24793>.

We rely on opam to test, every night, if our packages still work with the latest git version of their dependencies. That use-case is entirely broken by the fact that opam does not update things with a `opam upgrade -a $PACKAGE` unless the package is explicitly listed.

**Comment by rjbou** [comment]
In this last case, I don't see any `working-dir` or dirty source message. According logs, `coq-stdpp` seems to be pinned ansdsynchronised on its dev vc `git+https://gitlab.mpi-sws.org/iris/stdpp.git#master`. In this case it should be upgraded  without need to add it as update argument.

**Comment by RalfJung** [comment]
Well, but it has not been upgraded: `gset_to_gmap` exists in std++ master, it was added by [this commit](https://gitlab.mpi-sws.org/iris/stdpp/commit/31e0d1f6f7d11cbec5476a357b1bdaf25857688d).

Adding it to the list helped (our CI scripts [now do that automatically](https://gitlab.mpi-sws.org/iris/ci/commit/26edd1e854954cad37637fd8e27fb5cffbc37841)).

**Comment by rjbou** [comment]
I don't think it is an upgrade command problem*, but environment problem, maybe not even related to just the upgrade of `coq-stdpp`. In and before this [log](https://gitlab.mpi-sws.org/iris/lambda-rust/-/jobs/24138), upgrade recompile on each upstream change, in this [one](https://gitlab.mpi-sws.org/iris/lambda-rust/-/jobs/24285), there is a version conflict (du to build-dep generation?), and since, coq-stdpp is no more updated ([this](https://gitlab.mpi-sws.org/iris/lambda-rust/-/jobs/24477), and [this](https://gitlab.mpi-sws.org/iris/lambda-rust/-/jobs/24693)).
Maybe a hint is to check `coq-lambda-rust-builddep` opam file, if there is a diff between `opam show --raw coq-lambda-rust-builddep` and `/builds/iris/lambda-rust/build-dep/opam`?

(*) Checked scenario `pin`-`update`-`pin`-`upgrade`, and as upstream changes, it is always updated (unless not installed, but it is note the case f `coq-stddp` if I follow the log/cache)

**Comment by RalfJung** [comment]
> in this [one](https://gitlab.mpi-sws.org/iris/lambda-rust/-/jobs/24285), there is a version conflict (du to build-dep generation?)

That conflict is just there being a newer version available than what lambda-rust uses. We want to get the latest versions *satisfying our constraints*, and from what I can tell opam just says here that there'd be newer versions but there are constraints blocking them from being installed. That seems fine.

**Comment by AltGr** [comment]
I just reproduced a similar scenario, which indeed would point to a bug (different from original one, but probably the one you hit last). It happened when changing a pinning target:
```
[lg@agaric{4.06.0} ~/ocamlpro/opam]31% opam pin z3 git+file:///home/lg/ocamlpro/z3#louis   
[NOTE] Package z3 is currently pinned to git+file:///home/lg/ocamlpro/z3#z3-4.8.4 (version 4.8.4).
[z3.4.8.4] synchronised from git+file:///home/lg/ocamlpro/z3#louis
z3 is now pinned to git+file:///home/lg/ocamlpro/z3#louis (version 4.8.4)

Everything as up-to-date as possible (run with --verbose to show unavailable upgrades).

The following packages are not being upgraded because the new versions conflict with other installed packages:
  - angstrom.0.11.0
  - [...]

However, you may "opam upgrade" these packages explicitly, which will ask permission to downgrade or uninstall the conflicting packages.
Nothing to do.
[lg@agaric{4.06.0} ~/ocamlpro/opam]% opam upgrade z3                                  

<><> Synchronising pinned packages ><><><><><><><><><><><><><><><><><><><><><><>
[z3.4.8.4] no changes from git+file:///home/lg/ocamlpro/z3#louis

The following actions will be performed:
  ↻ recompile z3 4.8.4*
[z3.4.8.4] no changes from git+file:///home/lg/ocamlpro/z3#louis
```

it's quite weird that the initial `opam pin` got `Nothing to do.`, while the following `opam upgrade z3` properly reinstalls it.
This means that the `opam pin` command correctly marked the package for reinstall, but failed to take it into account directly. Maybe a missing `request_packages` argument at some point ?

**Comment by rjbou** [comment]
What is the status of this issue, given referenced PRs included in 2.0.4?

**Comment by rjbou** [comment]
Closing this issues, feel free to reopen it if needed.
Thanks!
